### PR TITLE
store/tikv: use grpc

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR replaces the protocol interaction with PD/TiKV to GRPC.

PTAL @coocood @tiancaiamao @disksing @siddontang

### CODE REVIEW

File: store/tikv/client.go [hunk 1]
```diff
@@ -15,140 +15,232 @@
...
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
+	var label = rpcLabelKV
+	if req.Type == tikvrpc.CmdCop {
+		label = rpcLabelCop
 	}
+	defer func() { sendReqHistogram.WithLabelValues(label).Observe(time.Since(start).Seconds()) }()
```
Comment (siddontang) [inline store/tikv/client.go]: do we need this metric now? May the ecosystem contain this?

File: store/tikv/region_request.go [hunk 2]
```diff
@@ -146,27 +113,19 @@ func (s *RegionRequestSender) sendKVReqToRegion(bo *Backoffer, ctx *RPCContext,
...
-	req.Context = ctx.KVCtx
-	resp, err = s.client.SendCopReq(bo.ctx, ctx.Addr, req, timeout)
-	if err != nil {
-		if e := s.onSendFail(bo, ctx, err); e != nil {
-			return nil, false, errors.Trace(err)
-		}
-		return nil, true, nil
+func (s *RegionRequestSender) onSendFail(bo *Backoffer, ctx *RPCContext, err error) error {
+	// If it failed because the context is canceled, don't retry on this error.
+	if errors.Cause(err) == goctx.Canceled || grpc.Code(err) == codes.Canceled {
```
Comment (tiancaiamao) [inline store/tikv/region_request.go]: Add a TODO here.
error code maybe not Canceled, and test case also need update.

File: store/tikv/client_test.go [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// Copyright 2015 Reborndb Org. All Rights Reserved.
```
Comment (tiancaiamao) [inline store/tikv/client_test.go]: ???

**Comment by coocood** [comment]
LGTM

File: store/tikv/2pc.go [hunk 4]
```diff
@@ -229,6 +231,17 @@ func (c *twoPhaseCommitter) doActionOnKeys(bo *Backoffer, action twoPhaseCommitA
 	return errors.Trace(err)
 }
 
+// reserveStack reserves 4KB memory on the stack to avoid runtime.morestack, call it after new a goroutine if necessary.
+func reserveStack(dummy bool) {
+	var buf [4 << 10]byte
```
Comment (tiancaiamao) [inline store/tikv/2pc.go]: I suggest change it to 8K

File: store/tikv/client.go (hunk 1 above)
Comment (hhkbp2) [inline store/tikv/client.go]: The grpc metrics calculate the sum time between sending requests and receiving response.
The metric here is to record the time it pushes request to grpc layer.

File: store/tikv/region_request.go (hunk 2 above)
Comment (hhkbp2) [inline store/tikv/region_request.go]: TODO tap should be made in that test case.

**Comment by hhkbp2** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**